contracts/builder-subnets/BuilderSubnets.sol

```diff
@@ -0,0 +1,502 @@
+        require(block.timestamp > minAllowedWithdrawalTimestamp, "BS: user withdraw is locked");
+        uint256 staked_ = staker.staked - amount_;
+        require(staked_ >= subnet.minStake || staked_ == 0, "BS: min stake reached");
+
+        _updateStorage(subnetId_, stakerAddress_, staked_, staker.claimLockEnd);
+
+        (uint256 fee_, address treasuryAddress_) = _getProtocolFee(amount_, FEE_WITHDRAW_OPERATION);
+        if (fee_ > 0) {
+            IERC20(token).safeTransfer(treasuryAddress_, fee_);
+            amount_ -= fee_;
+
+            emit FeePaid(subnetId_, stakerAddress_, fee_, treasuryAddress_);
+        }
+
+        IERC20(token).safeTransfer(stakerAddress_, amount_);
+
+        emit Withdrawn(subnetId_, stakerAddress_, staker, amount_);
+    }
+
+    function claim(bytes32 subnetId_, address stakerAddress_) external onlyExistedSubnet(subnetId_) {
+        Staker storage staker = stakers[subnetId_][stakerAddress_];
+        require(block.timestamp > staker.claimLockEnd, "BS: claim is locked");
+
+        _updateStorage(subnetId_, stakerAddress_, staker.staked, staker.claimLockEnd);
+        uint256 toClaim_ = staker.pendingRewards;
+        staker.pendingRewards = 0;
+
+        require(toClaim_ > 0, "BS: nothing to claim");
+
+        (uint256 protocolFee_, address protocolTreasury_) = _getProtocolFee(toClaim_, FEE_CLAIM_OPERATION);
+        (uint256 subnetFee_, address subnetTreasury_) = _getSubnetFee(toClaim_, subnetId_);
+        if (protocolFee_ > 0) {
+            IERC20(token).safeTransferFrom(treasury, protocolTreasury_, protocolFee_);
+            toClaim_ -= protocolFee_;
+
+            emit FeePaid(subnetId_, stakerAddress_, protocolFee_, protocolTreasury_);
+        }
+        if (subnetFee_ > 0) {
+            if (subnetFee_ > toClaim_) {
+                subnetFee_ = toClaim_;
+            }
+            IERC20(token).safeTransferFrom(treasury, subnetTreasury_, subnetFee_);
+            toClaim_ -= subnetFee_;
+
+            emit FeePaid(subnetId_, stakerAddress_, subnetFee_, subnetTreasury_);
+        }
+        if (toClaim_ > 0) {
+            IERC20(token).safeTransferFrom(treasury, stakerAddress_, toClaim_);
+
+            emit Claimed(subnetId_, stakerAddress_, staker, toClaim_);
+        }
+    }
+
+    /**
+     * @dev With claiming, there can be so many calculation periods that a transaction
+     * won't fit into a block. In this case, we can use this function to calculate
+     * rewards in parts.
+     */
+    function collectPendingRewards(
+        bytes32 subnetId_,
+        address stakerAddress_,
+        uint128 to_
+    ) external onlyExistedSubnet(subnetId_) {
+        Staker storage staker = stakers[subnetId_][stakerAddress_];
+
+        to_ = to_ > block.timestamp ? uint128(block.timestamp) : to_;
+        uint256 pendingRewards_ = getStakerRewards(subnetId_, stakerAddress_, to_);
+
+        staker.lastInteraction = to_;
```

Comment:
We will update lastIntercation even if the user doesn't stake anything. Is it OK?